crates/c-api/src/func.rs

```diff
@@ -284,6 +329,113 @@ pub(crate) unsafe fn c_callback_to_rust_fn(
     }
 }
 
+#[cfg(feature = "async")]
+struct CHostCallFuture<'a> {
+    pub callback: wasmtime_func_async_continuation_callback_t,
+    pub env: crate::ForeignData,
+    pub caller: wasmtime_caller_t<'a>,
+    pub param_count: usize,
+    pub results: &'a mut [Val],
+    pub hostcall_val_storage: Vec<wasmtime_val_t>,
+}
```

Comment:
One thing that worries me about this is that there's an implicit contract that while this future is alive it cannot access `hostcall_val_storage` because that's been handed out to C and it is expected to retain pointers to it. Within the `poll` method, however, there's nothing stopping the impementation from accessing this variable.

I'm trying to think of how to best restructure this to best handle this, but nothing feels like a slam dunk to me. That being said, how would you feel about:

* Minmizing the size fo this `CHostCallFuture` to basically dealing with `wasmtime_async_continuation_t`. (e.g. you may be able to get away with `impl Future for wasmtime_async_continuation_t` or something like that.
* Moving the bulk of this function to an `async fn`. The `async fn` would take as arguments most of the struct fields here, which would guarantee they're valid for the lifetime of the function.
* Within the `async fn` there'd be a `.await` on `wasmtime_async_continuation_t` from above (or the equivalent thereof). This would prevent access to the local variables while the function is awaiting.

That feels enough to me to prevent footguns for now, despite it not being 100% bulletproof.

Reply:
Okay I took a stab at this - I wasn't able to figure out how to make `wasmtime_async_continuation_t` a future without remove the `caller` parameter, but maybe that should follow suit of the results and params args and just make the caller keep a reference to that too.

Reply:
Okay I did that and see what you were saying. Nice suggestion!